executor/executor.go

```diff
@@ -930,21 +930,47 @@ type SelectLockExec struct {
 	Lock *ast.SelectLockInfo
 	keys []kv.Key
 
+	// The children may be a join of multiple tables, so we need a map.
 	tblID2Handle map[int64][]plannercore.HandleCols
 
-	// All the partition tables in the children of this executor.
-	partitionedTable []table.PartitionedTable
+	// When SelectLock work on a partition table, we need the partition ID
+	// (Physical Table ID) instead of the 'logical' table ID to calculate
+	// the lock KV. In that case, the Physical Table ID is extracted
+	// from the row key in the store and as an extra column in the chunk row.
 
-	// When SelectLock work on the partition table, we need the partition ID
-	// instead of table ID to calculate the lock KV. In that case, partition ID is store as an
-	// extra column in the chunk row.
-	// tblID2PIDColumnIndex stores the column index in the chunk row. The children may be join
-	// of multiple tables, so the map struct is used.
-	tblID2PIDColumnIndex map[int64]int
+	// tblID2PhyTblIDCol is used for partitioned tables.
+	// The child executor need to return an extra column containing
+	// the Physical Table ID (i.e. from which partition the row came from)
+	// Used during building
+	tblID2PhysTblIDCol map[int64]*expression.Column
+
+	// Used during execution
+	// Map from logic tableID to column index where the physical table id is stored
+	// For dynamic prune mode, model.ExtraPhysTblID columns are requested from
+	// storage and used for physical table id
+	// For static prune mode, model.ExtraPhysTblID is still sent to storage/Protobuf
+	// but could be filled in by the partitions TableReaderExecutor
+	// due to issues with chunk handling between the TableReaderExecutor and the
+	// SelectReader result.
```

Comment:
Seems like the approach for dynamic-prune-mode can also work under static-prune-mode? Do we have a plan to merge them into one to make it easier to maintain?

Reply:
For the SelectLockExec implementation's use of `tblID2PhysTblIDColIdx` it is already the same for dynamic and static prune mode. Only for UnionScan there might be a difference since UnionScan will be done per partition while SelectLock will be done per table (with a PartitionUnion for static prune mode IIRC).

Reply:
Just clarifying:
The TableReader will not fill in PhysTblID/PidColID with the physical table id. For dynamic prune mode it will always be done in storage. For static prune mode it will be filled in by storage in the SelectLock case, and for UnionScan it only uses a single partition/physical table id per TableReader, so the us.table.RecordPrefix() can be used directly.